### PR TITLE
feat!: Support Multiple Syncs

### PR DESCRIPTION
**Pull request by james-milligan**
Signed-off-by: James Milligan <james@omnant.co.uk>

corresponding flagd changes [here](https://github.com/open-feature/flagd/pull/222)

## This PR

- updates the `openfeature.dev/featureflagconfiguration` to allow for multiple syncs to be provided to the flagd sidecar, the annotation value is now a comma separated list with each value providing both the namespace and the CRD name separated with a `/`, if no namespace is provided it is assumed that the crd is in the same namespace as the deployment e.g. `"end-to-end, test/end-to-end-2"`
- introduces remote sync implementation

### Related Issues

https://github.com/open-feature/open-feature-operator/issues/251

### How to test
- This is a breaking change and will not work with the existing flagd version, a build from [this PR](https://github.com/open-feature/flagd/pull/222) will be required to test

### CODE REVIEW

File: apis/core/v1alpha1/featureflagconfiguration_types.go
Comment (skyerus) [inline apis/core/v1alpha1/featureflagconfiguration_types.go:58]: We now support multiple versions of the crd (see the v1alpha2 folder). I suppose these will be breaking changes? If yes we would need to introduce a new version of the crd

File: config/crd/bases/core.openfeature.dev_featureflagconfigurations.yaml
Comment (toddbaert) [inline config/crd/bases/core.openfeature.dev_featureflagconfigurations.yaml:243]: Is http better than "remote"?  "Remote" seems a bit nonspecific.

**Comment by toddbaert** [comment]
is there a reason to go with a comma delimited string over a proper yaml array? Is it just for more backward compatibility? If so, maybe there's another way to achieve it?

File: webhooks/pod_webhook.go
Comment (toddbaert) [inline webhooks/pod_webhook.go:243]: it may be simpler to check `featureFlag.Spec.SyncProvider != nil` just once and use nested `if`s, but I'll leave that up to you.

**Comment by toddbaert** [comment]
I really like the `--uri` arguments, they may the flagd command line much more sensible IMHO, but I think we should proceed really carefully with the CRD spec. I'd like to try our best to maintain compatibility but also adhere to conventions (like yaml arrays) and make things easy to understand.

**Comment by james-milligan** [comment]
> is there a reason to go with a comma delimited string over a proper yaml array? Is it just for more backward compatibility? If so, maybe there's another way to achieve it?
> 
> 
> 
> Maybe we could add `openfeature.dev/featureflagconfigurations` (note the "s") which could be an array?
> 
> 
> 
> I think this would be backward compatible... is this a completely dumb suggestion? @james-milligan @skyerus @AlexsJones 

Unfortunately the annotation values must be of type string - I agree an array would be far nicer!

**Comment by toddbaert** [comment]
> > is there a reason to go with a comma delimited string over a proper yaml array? Is it just for more backward compatibility? If so, maybe there's another way to achieve it?
> > Maybe we could add `openfeature.dev/featureflagconfigurations` (note the "s") which could be an array?
> > I think this would be backward compatible... is this a completely dumb suggestion? @james-milligan @skyerus @AlexsJones
> 
> Unfortunately the annotation values must be of type string - I agree an array would be far nicer!

welp, that explains that. :sweat_smile:

File: test/e2e/e2e.yml
Comment (AlexsJones) [inline test/e2e/e2e.yml:62]: Can we use `/` as is standard in k8s? e.g. `namespace/service`

File: config/crd/bases/core.openfeature.dev_featureflagconfigurations.yaml
Comment (james-milligan) [reply config/crd/bases/core.openfeature.dev_featureflagconfigurations.yaml:243]: will update this to `http`

**Comment by james-milligan** [comment]
> Maybe we could add openfeature.dev/featureflagconfigurations (note the "s") which could be an array?

ill update the parsing to assume that the current requests namespace is to be used if none is defined, this will keep it backwards compatible (excluding the CRD changes required to enable the remote/http sync)

File: webhooks/pod_webhook.go
Comment (james-milligan) [reply webhooks/pod_webhook.go]: opted for a switch instead

File: test/e2e/e2e.yml
Comment (james-milligan) [reply test/e2e/e2e.yml]: ive introduced this change in both PRs

**Comment by james-milligan** [comment]
e2e tests will continue to fail until the corresponding PR in flagd is merged, but I have tested it locally

File: README.md
Comment (AlexsJones) [inline README.md:114]: Is it necessary?

File: webhooks/pod_webhook.go
Comment (AlexsJones) [inline webhooks/pod_webhook.go:264]: Is bearer-token now mandatory or are we passing a nil value here?

File: webhooks/pod_webhook.go
Comment (AlexsJones) [inline webhooks/pod_webhook.go:276]: Perhaps this should be a constant path?

File: webhooks/pod_webhook.go
Comment (AlexsJones) [inline webhooks/pod_webhook.go:290]: Again, perhaps this should be constant

File: webhooks/pod_webhook.go
Comment (AlexsJones) [inline webhooks/pod_webhook.go:96]: Here we are saying that if the provider is filepath or http we should create the configmap; does http need it?

File: README.md
Comment (james-milligan) [reply README.md]: removed

File: webhooks/pod_webhook.go
Comment (james-milligan) [reply webhooks/pod_webhook.go]: I've now wrapped this in an `if`

File: webhooks/pod_webhook.go
Comment (james-milligan) [reply webhooks/pod_webhook.go]: added constant for file path

File: webhooks/pod_webhook.go
Comment (james-milligan) [reply webhooks/pod_webhook.go]: updated this to check for file path sync, rather than non-kubernetes sync

File: webhooks/pod_webhook.go
Comment (skyerus) [inline webhooks/pod_webhook.go:27]:
```suggestion
	fileSyncMountPath      string            = "/etc/flagd/"
```

**Review by AlexsJones — APPROVED** [review]
Nice work young man.
Please make sure this doesn't break the demo/watchman

File: apis/core/v1alpha2/featureflagconfiguration_conversion.go
Comment (skyerus) [inline apis/core/v1alpha2/featureflagconfiguration_conversion.go:75]: Do we not need similar logic in the `ConvertTo` func?

**Review by skyerus — APPROVED** [review]

**Pull request merged**
